runtime/io/io.cpp

```diff
@@ -0,0 +1,389 @@
+    case ECONNABORTED:
+      retBlock->h = getBlockHeaderForSymbol((((uint64_t)getTagForSymbolName(ERRTAG(ECONNABORTED))) << 32) | 1);
+    case ECONNRESET:
+      retBlock->h = getBlockHeaderForSymbol((((uint64_t)getTagForSymbolName(ERRTAG(ECONNRESET))) << 32) | 1);
+    case ENOBUFS:
+      retBlock->h = getBlockHeaderForSymbol((((uint64_t)getTagForSymbolName(ERRTAG(ENOBUFS))) << 32) | 1);
+    case EISCONN:
+      retBlock->h = getBlockHeaderForSymbol((((uint64_t)getTagForSymbolName(ERRTAG(EISCONN))) << 32) | 1);
+    case ENOTCONN:
+      retBlock->h = getBlockHeaderForSymbol((((uint64_t)getTagForSymbolName(ERRTAG(ENOTCONN))) << 32) | 1);
+    case ESHUTDOWN:
+      retBlock->h = getBlockHeaderForSymbol((((uint64_t)getTagForSymbolName(ERRTAG(ESHUTDOWN))) << 32) | 1);
+    case ETOOMANYREFS:
+      retBlock->h = getBlockHeaderForSymbol((((uint64_t)getTagForSymbolName(ERRTAG(ETOOMANYREFS))) << 32) | 1);
+    case ETIMEDOUT:
+      retBlock->h = getBlockHeaderForSymbol((((uint64_t)getTagForSymbolName(ERRTAG(ETIMEDOUT))) << 32) | 1);
+    case ECONNREFUSED:
+      retBlock->h = getBlockHeaderForSymbol((((uint64_t)getTagForSymbolName(ERRTAG(ECONNREFUSED))) << 32) | 1);
+    case EHOSTDOWN:
+      retBlock->h = getBlockHeaderForSymbol((((uint64_t)getTagForSymbolName(ERRTAG(EHOSTDOWN))) << 32) | 1);
+    case EHOSTUNREACH:
+      retBlock->h = getBlockHeaderForSymbol((((uint64_t)getTagForSymbolName(ERRTAG(EHOSTUNREACH))) << 32) | 1);
+    case ELOOP:
+      retBlock->h = getBlockHeaderForSymbol((((uint64_t)getTagForSymbolName(ERRTAG(ELOOP))) << 32) | 1);
+    case EOVERFLOW:
+      retBlock->h = getBlockHeaderForSymbol((((uint64_t)getTagForSymbolName(ERRTAG(EOVERFLOW))) << 32) | 1);
+    default:
+      retBlock = static_cast<block *>(koreResizeLastAlloc(retBlock, sizeof(block) + sizeof(uint64_t), sizeof(block)));
+      retBlock->h = getBlockHeaderForSymbol((uint64_t)getTagForSymbolName("Lbl'Hash'unkownIOError"));
+      mpz_t err;
+      mpz_init_set_si(err, errno);
+      mpz_ptr p = move_int(err);
+      memcpy(&(retBlock->children[0]), &p, sizeof(uint64_t));
+    }
+
+    return retBlock;
+  }
+
+  blockheader header_int() {
+    static blockheader header = {(uint64_t)-1};
+
+    if (header.hdr == -1) {
+      getBlockHeaderForSymbol((uint64_t)getTagForSymbolName("inj{SortInt, SortIOInt}"));
+    }
+
+    return header;
+  }
+
+  blockheader header_err() {
+    static blockheader header = {(uint64_t)-1};
+
+    if (header.hdr == -1) {
+      getBlockHeaderForSymbol((uint64_t)getTagForSymbolName("inj{SortIOError, SortKItem}"));
+    }
+
+    return header;
+  }
+
+  blockheader header_string() {
+    static blockheader header = {(uint64_t)-1};
+
+    if (header.hdr == -1) {
+      getBlockHeaderForSymbol((uint64_t)getTagForSymbolName("inj{SortString, SortIOString}"));
+    }
+
+    return header;
+  }
+
+  char * getTerminatedString(string * str) {
+    int length = len(str);
+    char * buf = (char *) malloc(sizeof(KCHAR) * (length + 1));
```

Comment:
can we make this a koreAllocToken instead? then you don't need to explicitly free and you aren't paying the overhead of a malloc. You will have to call set_len though.